---
output: github_document
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
# Please put your title here to include it in the file below.
Title <- "Single cell resolution landscape of equine peripheral blood mononuclear cells reveals diverse cell types including T-bet<sup>+</sup> B cells."
```

|                                                                             ![epbmclogo](analysis/data/derived_data/horse_DNA_image.png#style=centerme)                                                   

# Single cell resolution landscape of equine peripheral blood mononuclear cells reveals diverse cell types including T-bet<sup>+</sup> B cells.

This repository contains the data analysis scripts and code for our paper:

> Roosheel S. Patel*, Joy E. Tomlinson*, Thomas J. Divers, Gerlinde R. Van de Walle, Brad R. Rosenberg, (2020). *Single cell resolution landscape of equine peripheral blood mononuclear cells reveals diverse cell types including T-bet+ B cells*. BMC Biology <https://doi.org/XXXX>

Our pre-print is online here:

> Roosheel S. Patel*, Joy E. Tomlinson*, Thomas J. Divers, Gerlinde R. Van de Walle, Brad R. Rosenberg, (2020). *Single cell resolution landscape of equine peripheral blood mononuclear cells reveals diverse cell types including T-bet+ B cells* biorxiv, Accessed `r format(Sys.Date(), "%d %b %Y")`. Online at <https://doi.org/10.1101/2020.05.05.077362>

## Contents

The **analysis** directory contains:

  - [:file\_folder: paper](/analysis/paper): R Markdown document which includes code to reproduce the figures generated by our analysis.
  - [:file\_folder: data](/analysis/data): Supplementary data files and metadata used in this analysis

## Additional files required for analysis

For surface annotation of differentially expressed genes, surface protein annotations were pulled from the [Cell Surface Protein Atlas website](https://wlab.ethz.ch/cspa/#abstract). 

> Bausch-Fluck D, Hofmann A, Bock T, Frei AP, Cerciello F, et al. (2015) *A Mass Spectrometric-Derived Cell Surface Protein Atlas*. PLoS One <https://doi.org/10.1371/journal.pone.0121314>

## Licenses

**Code :** MIT License, Copyright (c) 2020 Roosheel Patel

The organization of this repository was adapted from the rrtools research compendium structure

> Marwick, B. (2017). Computational reproducibility in archaeological research: Basic principles and a case study of their implementation.*Journal of Archaeological Method and Theory*, 24(2), 424-450.<https://doi.org/10.1007/s10816-015-9272-9>

## Session Information

Session information (computer environment and R/package versions) used to generate the analyses and figures in the manuscript:

<details>
>R version 3.6.2 (2019-12-12)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)
>Matrix products: default
BLAS/LAPACK: /hpc/packages/minerva-centos7/intel/parallel_studio_xe_2019/compilers_and_libraries_2019.0.117/linux/mkl/lib/intel64_lin/libmkl_gf_lp64.so
>Random number generation:
 RNG:     Mersenne-Twister 
 Normal:  Inversion 
 Sample:  Rounding 
>locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C     
>other attached packages:
 [1] readr_1.4.0          modelr_0.1.8         tidyr_1.1.2          reshape2_1.4.4       here_1.0.0           circlize_0.4.11      ComplexHeatmap_2.2.0
 [8] GEOquery_2.54.1      factoextra_1.0.7     cluster_2.1.0        openxlsx_4.2.3       legocolors_0.2.0     ggforce_0.3.2        ggtree_2.0.4        
[15] viridis_0.5.1        viridisLite_0.3.0    BiocManager_1.30.10  jntools_0.1.0        mgsub_1.7.2          ape_5.4-1            patchwork_1.1.0     
[22] dplyr_1.0.2          edgeR_3.28.1         limma_3.42.2         scales_1.1.1         lemon_0.4.5          ggpubr_0.4.0         readxl_1.3.1        
[29] ggrepel_0.8.2        RColorBrewer_1.1-2   ggnetwork_0.5.8      sctransform_0.2.0    uwot_0.1.9           rlang_0.4.9          rrtools_0.1.0       
[36] future_1.20.1        clustree_0.4.3       ggraph_2.0.4         ggplot2_3.3.2        gdata_2.18.0         cowplot_1.1.0        gtools_3.8.2        
[43] tibble_3.0.4         AnnotationDbi_1.48.0 IRanges_2.20.2       S4Vectors_0.24.4     Biobase_2.46.0       BiocGenerics_0.32.0  Matrix_1.2-18       
[50] renv_0.12.3          Seurat_3.1.0        
>loaded via a namespace (and not attached):
  [1] reticulate_1.18       tidyselect_1.1.0      RSQLite_2.2.1         htmlwidgets_1.5.2     Rtsne_0.15            devtools_2.3.2       
  [7] munsell_0.5.0         codetools_0.2-18      ica_1.0-2             miniUI_0.1.1.1        withr_2.3.0           colorspace_2.0-0     
 [13] knitr_1.30            rstudioapi_0.13       ROCR_1.0-11           ggsignif_0.6.0        tensor_1.5            listenv_0.8.0        
 [19] labeling_0.4.2        git2r_0.27.1          polyclip_1.10-0       bit64_4.0.5           farver_2.0.3          rprojroot_2.0.2      
 [25] treeio_1.10.0         parallelly_1.21.0     vctrs_0.3.5           generics_0.1.0        xfun_0.19             R6_2.5.0             
 [31] clue_0.3-58           graphlayouts_0.7.1    rsvd_1.0.3            locfit_1.5-9.4        spatstat.utils_1.17-0 assertthat_0.2.1     
 [37] promises_1.1.1        gtable_0.3.0          globals_0.14.0        processx_3.4.5        goftest_1.2-2         tidygraph_1.2.0      
 [43] clisymbols_1.2.0      GlobalOptions_0.1.2   splines_3.6.1         rstatix_0.6.0         lazyeval_0.2.2        broom_0.7.2          
 [49] yaml_2.2.1            abind_1.4-5           backports_1.2.1       httpuv_1.5.4          tools_3.6.1           usethis_1.6.3        
 [55] bookdown_0.21         ellipsis_0.3.1        sessioninfo_1.1.1     ggridges_0.5.2        Rcpp_1.0.5            plyr_1.8.6           
 [61] purrr_0.3.4           ps_1.5.0              prettyunits_1.1.1     rpart_4.1-15          deldir_0.2-3          GetoptLong_1.0.4     
 [67] pbapply_1.4-3         zoo_1.8-8             haven_2.3.1           fs_1.5.0              magrittr_2.0.1        data.table_1.13.4    
 [73] lmtest_0.9-38         RANN_2.6.1            fitdistrplus_1.1-3    matrixStats_0.57.0    pkgload_1.1.0         hms_0.5.3            
 [79] mime_0.9              evaluate_0.14         xtable_1.8-4          rio_0.5.16            shape_1.4.5           gridExtra_2.3        
 [85] testthat_3.0.0        compiler_3.6.1        KernSmooth_2.23-18    crayon_1.3.4          htmltools_0.5.0       mgcv_1.8-33          
 [91] later_1.1.0.1         DBI_1.1.0             tweenr_1.0.1          MASS_7.3-53           car_3.0-10            cli_2.2.0            
 [97] igraph_1.2.6          forcats_0.5.0         pkgconfig_2.0.3       rvcheck_0.1.8         foreign_0.8-71        plotly_4.9.2.1       
[103] xml2_1.3.2            stringr_1.4.0         callr_3.5.1           digest_0.6.27         RcppAnnoy_0.0.17      spatstat.data_1.5-2  
[109] rmarkdown_2.5         cellranger_1.1.0      leiden_0.3.6          tidytree_0.3.3        curl_4.3              shiny_1.5.0          
[115] rjson_0.2.20          lifecycle_0.2.0       nlme_3.1-150          jsonlite_1.7.1        carData_3.0-4         desc_1.2.0           
[121] fansi_0.4.1           pillar_1.4.7          lattice_0.20-41       fastmap_1.0.1         httr_1.4.2            pkgbuild_1.1.0       
[127] survival_3.2-7        glue_1.4.2            remotes_2.2.0         zip_2.1.1             spatstat_1.64-1       png_0.1-7            
[133] bit_4.0.4             stringi_1.5.3         blob_1.2.1            memoise_1.1.0         irlba_2.3.3           future.apply_1.6.0   
